compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -403,12 +406,35 @@ public void visit(BLangResourceFunction funcNode, AnalyzerData data) {
             BLangType restParamType = funcNode.resourcePathType.restParamType;
             symResolver.resolveTypeNode(restParamType, data.env);
             if (!types.isAssignable(restParamType.getBType(), symTable.pathParamAllowedType)) {
-                dlog.error(restParamType.getPosition(), DiagnosticErrorCode.UNSUPPORTED_REST_PATH_PARAM_TYPE, 
+                dlog.error(restParamType.getPosition(), DiagnosticErrorCode.UNSUPPORTED_REST_PATH_PARAM_TYPE,
                         restParamType.getBType());
             }
         }
     }
 
+    private void checkForClientObjectTypeOrFunctionType(BType type, BType referredType, Location pos) {
+        if (referredType.tsymbol != null && Symbols.isFlagOn(referredType.tsymbol.flags, Flags.CLIENT)) {
+            dlog.error(pos, DiagnosticErrorCode.INVALID_RESOURCE_METHOD_RETURN_TYPE, "client object", type);
+        }
+        switch (referredType.tag) {
+            case TypeTags.INVOKABLE:
+                dlog.error(pos, DiagnosticErrorCode.INVALID_RESOURCE_METHOD_RETURN_TYPE, "function", type);
+                break;
+            case TypeTags.UNION:
+                for (BType memberType: ((BUnionType) referredType).getMemberTypes()) {
+                    checkForClientObjectTypeOrFunctionType(type, memberType, pos);
+                }
```

Comment:
```ballerina
type CustomType0 float|int[]|client object{}|int|string;
type CustomType1 float|int[]|client object{}|int|string;
type CustomType3 CustomType0|CustomType1;
type CustomType4 CustomType3;

client object{} & readonly successClient = client object {

};

public function testErrorInResourceMethodReturningObjectType() {
    var successClientObject = client object {
        resource function get clientReturnPath() returns CustomType4 {
            return successClient;
        }
    };
}
```
For the above example, we get the same error twice. 
Error
```
ERROR [test3.bal:(9:58,9:69)] return type of the resource function does not allow 'client object' type: 'CustomType4'
ERROR [test3.bal:(9:58,9:69)] return type of the resource function does not allow 'client object' type: 'CustomType4'

```

Reply:
Solved 2725ba33d9230d4c98b3b79818080c3ba31e21b3